### ISSUE TITLE
Allowing string for comparison operator >,<,>=,<=

### ISSUE THREAD

**Issue by patrick-yi-82** [yiisoft/yii2#16034]
DataFilter doesn't allow string for comparison operator >,<,>=,<= by default.
However mysql, mssql and php all allows string comparisons.
ex : 'ball' < 'water' is true
Shouldn't DataFilter allow it too?

```
class DataFilter extends Model
{

    public $operatorTypes = [
        '<' => [self::TYPE_INTEGER, self::TYPE_FLOAT, self::TYPE_DATETIME, self::TYPE_DATE, self::TYPE_TIME,],
        '>' => [self::TYPE_INTEGER, self::TYPE_FLOAT, self::TYPE_DATETIME, self::TYPE_DATE, self::TYPE_TIME],
        '<=' => [self::TYPE_INTEGER, self::TYPE_FLOAT, self::TYPE_DATETIME, self::TYPE_DATE, self::TYPE_TIME],
        '>=' => [self::TYPE_INTEGER, self::TYPE_FLOAT, self::TYPE_DATETIME, self::TYPE_DATE, self::TYPE_TIME],
        '=' => '*',
        '!=' => '*',
        'IN' => '*',
        'NOT IN' => '*',
        'LIKE' => [self::TYPE_STRING],
    ];

    public function init()
    {
        // Manually adding string to comparison operators
        $this->operatorTypes['<'][] = self::TYPE_STRING;
        $this->operatorTypes['>'][] = self::TYPE_STRING;
        $this->operatorTypes['<='][] = self::TYPE_STRING;
        $this->operatorTypes['>='][] = self::TYPE_STRING;

        parent::init();
    }

```

### What do you get instead?
`[{"field":"filter","message":"\"someString\" does not support operator \"lt\"."}]`

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.14
| PHP version      | 
| Operating system |

**Comment by itma** [comment]
I think it is a db dependent problem, and possible implementation of it requires configuring DataFilter before using it, for example, by reading the app configuration database connection (can be few, which to choose?), then inject the properer DataFilterDriver (or something like that) particular for each database containg a set of allowed operators.

Better doing it on db side by a function. Nevertheless I can be wrong :)

**Comment by patrick-yi-82** [comment]
Indeed, I was assuming it might be a db dependent feature. But to my knowledge, most modern db and programming languages allows string comparison, >,<,<= and >=.

**Comment by samdark** [comment]
It is like that because mostly these operands do not make sense with strings if we're talking about data filtering. Property is public so if you need it you can always configure it.

**Comment by rob006** [comment]
> It is like that because mostly these operands do not make sense with strings if we're talking about data filtering.

It makes perfect sense. If I want to find records starting from `o` to `z` all I need is to use `>=o`. I've used this feature multiple times in CRUDs.

**Comment by samdark** [comment]
OK. Let's fix it then.
